### PR TITLE
231221 support multi project and repo

### PR DESCRIPTION
**Pull request by amerharb**
### Added

- Server config file `./config/projects.yaml` to support multi project and multi repository
- New API endpoint to get all projects in a list that is configured at server in `./config/projects.yaml`
  ```http request
  GET http://example.com/api/projects
  ```
- New page to list all projects and languages
  ```
  http://example.com/projects
  ```

### Changed

- Api url to include project name in path
  ```http request
  GET http://example.com/api/translations/<projectName>/sv
  ```
- Web page url to include which project to load
  ```
  http://example.com/projects/<projectName>/sv

### CODE REVIEW

**Comment by richardolsson** [comment]
I notice this PR is still in draft mode. Is it ready for review?

**Comment by richardolsson** [comment]
I'm going to assume it is ready for review since @amerharb requested a review from me.

**Comment by richardolsson** [comment]
I'm unable to run this because I can't figure out where to put the config file. Where should it go, and what should it be called?

**Comment by richardolsson** [comment]
Never mind, I managed to get it working by putting `projects.yaml` in `config` in the repo root, i.e. next to (rather than inside) the `webapp` directory.

File: webapp/src/store/Store.spec.ts
Comment (richardolsson) [inline webapp/src/store/Store.spec.ts:6]: I think these should be separated in two separate files, one class test suite per file.

File: webapp/src/store/Store.ts
Comment (richardolsson) [inline webapp/src/store/Store.ts:9]: I think this new class should have been in a separate file from the old one.

File: webapp/src/utils/config.spec.ts
Comment (richardolsson) [inline webapp/src/utils/config.spec.ts:12]: Same thing here. I don't like it when changes like these are made easily. Indenting an entire file is not a minor thing, and I'm honestly not sure I understand the need to have a `describe('config.ts')`. What information does that provide about the code, other than the file the tested code is in (which is already revealed by the name of the test file).

If you want to test two classes in a single test file, you can just have two top-level `describe()`, right?

But ultimately, I think `ServerConfig` should be in a separate file.

**Review by richardolsson — COMMENTED** [review]
Here are some initial comments following up on our zoom call today. Nice work on all of this!

I think we need to improve our process a little bit in order to not change so many things in a single PR, and communicate more before taking on big tasks like this one. In this case I think this task could have been simplified a by doing some refactors before the task was taken on.

But for now, let's do the bare minimum to get this merged and we can do such refactors afterwards instead.

File: webapp/src/Cache.ts
Comment (amerharb) [reply webapp/src/Cache.ts:67]: Good catch

File: webapp/src/utils/config.spec.ts
Comment (amerharb) [reply webapp/src/utils/config.spec.ts:12]: usually for me the point of having one describe that wrap all file and named after the name are:
1. For some IDE like WebStorm it makes it easy by one click to run all unit test in this file
2. AFAIK there is no command in Jest to run a test file but it only can be called by it describe

in all cases this will be split into 2 files as you mentioned above to follow splitting of config.ts

File: webapp/src/store/Store.ts
Comment (amerharb) [reply webapp/src/store/Store.ts:9]: will do

File: webapp/src/store/Store.spec.ts
Comment (amerharb) [reply webapp/src/store/Store.spec.ts:6]: will do

**Comment by amerharb** [comment]
> Never mind, I managed to get it working by putting `projects.yaml` in `config` in the repo root, i.e. next to (rather than inside) the `webapp` directory.

@richardolsson do you think it is better to move it to package root `webapp` instead of repo root?

File: webapp/src/utils/config.spec.ts
Comment (amerharb) [reply webapp/src/utils/config.spec.ts]: done seperate

File: webapp/src/store/Store.spec.ts
Comment (amerharb) [reply webapp/src/store/Store.spec.ts]: done

File: webapp/src/Cache.ts
Comment (amerharb) [reply webapp/src/Cache.ts]: Done, for this I did manual test locally but for later I have added 3 todos unit test to write them later in seprate PR

**Review by WULCAN — CHANGES_REQUESTED** [review]
I have reviewed all changes in this merge request up until e7fc5da5166119344a0c2b774e1fc309b049258b.

There are a lot of changes so it took a lot of time to review but reviewing is a lot less work than designing and implementing a new feature like this. Thank you very much for all your work in this merge request, @amerharb !

The changes enable support for repositories with multiple projects, as well as using a single Lyra instance with multiple repositories and I do not think there are any severe problems with these changes.

I do have one question, a few important observations, some ideas for future improvements, some findings on missing test coverage and some other minor findings. See my other comments in this review.

File: webapp/src/types.ts
Comment (amerharb) [reply webapp/src/types.ts:18]: done

File: webapp/src/utils/lyraConfig.spec.ts
Comment (amerharb) [reply webapp/src/utils/lyraConfig.spec.ts:184]: done

File: webapp/src/utils/config.spec.ts
Comment (amerharb) [reply webapp/src/utils/config.spec.ts:120]: in cases name promise is bad, renamed to actual and remove arrow function not really needed, deal with promise directly

File: webapp/src/utils/config.spec.ts
Comment (amerharb) [reply webapp/src/utils/config.spec.ts:183]: done

File: webapp/CHANGELOG.md
Comment (amerharb) [reply webapp/CHANGELOG.md]: yes it should be in release date, I will change this to a current date, 
however I talked with Richard about it, we will not continue update CHANGELOG at least with current phase or setup

File: webapp/src/utils/config.spec.ts
Comment (WULCAN) [reply webapp/src/utils/config.spec.ts]: This is resolved on 4217140a464f5558e9c0a7e1754a7d369126a7a0 where the constants contain the promise directly and are named actual.

File: webapp/src/utils/config.spec.ts
Comment (WULCAN) [reply webapp/src/utils/config.spec.ts]: On 4217140a464f5558e9c0a7e1754a7d369126a7a0, the desribe-block describes ProjectPathNotFoundError instead. This makes more sense to me.

File: webapp/src/utils/config.spec.ts
Comment (WULCAN) [reply webapp/src/utils/config.spec.ts]: On 4217140a464f5558e9c0a7e1754a7d369126a7a0, this has been resolved by storing the returned promise in actual instead of storing a function producing the promise.

File: webapp/src/app/api/translations/[projectName]/[lang]/[msgId]/route.ts
Comment (WULCAN) [reply webapp/src/app/api/translations/[projectName]/[lang]/[msgId]/route.ts:42]: Yes, I think it should be a 500 instead. The user requested something by project name. A project with that name was found. If the name was incorrect, we would have caught PorjectNameNotFoundError.

While 404 is for when the resource was not found, and that fits this situation, all 4XX codes are for errors that seem to have been caused by the client.

It is a very small problem so if we want to fix it in a later merge request, we can do so even without adding a TODO for it in the source code with this merge request.

File: webapp/README.md
Comment (WULCAN) [reply webapp/README.md:7]: Actually, the previous line still refers to an environment variable so we should fix the whole sentence, not just this last line.

File: webapp/src/Cache.ts
Comment (amerharb) [reply webapp/src/Cache.ts:32]: move the guard to the inside of function gitPull and rename it to gitPUllIfNeeded

File: webapp/src/app/api/translations/[lang]/[msgId]/route.ts
Comment (amerharb) [reply webapp/src/app/api/translations/[lang]/[msgId]/route.ts]: good catch, I will fix it

Comment (amerharb) [reply webapp/src/app/api/translations/[lang]/[msgId]/route.ts]: included lyraConfig.getProjectConfigByPath in the try catch, and add TODO for other calls to be done later that should return a server side  500 response

File: webapp/src/store/Store.spec.ts
Comment (amerharb) [reply webapp/src/store/Store.spec.ts]: After talked about it we decide to it as long as there is an async call

File: webapp/src/store/Store.spec.ts
Comment (WULCAN) [reply webapp/src/store/Store.spec.ts]: Ok, but then I want a test that verifies that `getProjectStore` after `updateTranslation` returns an updated `ProjectStore`. With just a 'get ProjectStore for a path' that uses `toEqual`, the implementation could return a clone of ProjectStore. No test would fail but `translations/.../[msgId]` would not work.

File: webapp/CHANGELOG.md
Comment (WULCAN) [reply webapp/CHANGELOG.md]: On 4217140a464f5558e9c0a7e1754a7d369126a7a0, this has been resolved by removing the date for 0.3.0

File: webapp/src/store/Store.ts
Comment (WULCAN) [reply webapp/src/store/Store.ts:16]: This is resolved on 4217140a464f5558e9c0a7e1754a7d369126a7a0

File: webapp/src/utils/config.ts
Comment (amerharb) [reply webapp/src/utils/config.ts:62]: now all paths in LyraConfig are abs

File: webapp/src/utils/lyraConfig.spec.ts
Comment (WULCAN) [reply webapp/src/utils/lyraConfig.spec.ts]: On 4217140a464f5558e9c0a7e1754a7d369126a7a0, both uses have been removed.

File: webapp/src/types.ts
Comment (WULCAN) [reply webapp/src/types.ts:18]: On 4217140a464f5558e9c0a7e1754a7d369126a7a0, `ProjectsResponse` is used in both files. `ProjectItem` is no longer explicitly used in `page.tsx` but still provides typing for the content of `ProjectResponse` there.

File: webapp/src/Cache.ts
Comment (WULCAN) [reply webapp/src/Cache.ts:32]: This has been resolved in 336f713e8e31d91a5fcc4a6efb606843922d88f5

File: webapp/src/app/api/translations/[lang]/[msgId]/route.ts
Comment (WULCAN) [reply webapp/src/app/api/translations/[lang]/[msgId]/route.ts]: Cool, I didn't realize the try block did not have anything actually throwing `ProjectPathNotFound`. With afdf580a4dcf2890d18db838ecf7f2ddf9451bf1, that is fixed but there is still nothing trowing `ProjectNameNotFoundError` in the try-block. dd412385a81726c2d7c6f6aad241d385845576a3 adds a TODO that covers that remaining issue.

File: webapp/src/utils/lyraConfig.ts
Comment (WULCAN) [inline webapp/src/utils/lyraConfig.ts]: `relativePath` is normalized, but `projectPath` is not. This will fail for a `projects.yaml` with multiple slashes or with dot elements.

normalization is probably the right thing to do, but then we have to do it in ServerConfig too.

File: webapp/src/utils/config.spec.ts
Comment (amerharb) [reply webapp/src/utils/config.spec.ts]: done

File: webapp/src/utils/config.ts
Comment (WULCAN) [reply webapp/src/utils/config.ts:62]: On 180f467e9fe9968a78ef01f10d9ab4d7a098f800, `path` is relative to `repoPath` and the two other paths are relative to `path`. All the paths are private now, which makes it much easier to review and understand to what reference they are relative to. I consider this resolved but note that there are still many different kinds of paths in Lyra source code, relative to different things.

File: webapp/src/store/Store.ts
Comment (amerharb) [reply webapp/src/store/Store.ts:4]: @WULCAN With reference to our discussion, this we be keep as is since project in Store is a bit diffreant than Project in LyraConfig.

File: webapp/src/utils/config.spec.ts
Comment (amerharb) [reply webapp/src/utils/config.spec.ts]: we talked about this

File: webapp/src/Cache.ts
Comment (amerharb) [reply webapp/src/Cache.ts]: fixed

File: webapp/src/utils/config.spec.ts
Comment (WULCAN) [reply webapp/src/utils/config.spec.ts]: This is resolved on c6c3ebc13694836e2cc4202277fe4cbcfad9e620 with a new test.

File: webapp/src/utils/config.ts
Comment (amerharb) [reply webapp/src/utils/config.ts]: add this suggestion to TODO text

File: webapp/src/Cache.ts
Comment (WULCAN) [reply webapp/src/Cache.ts]: This is Resolved on 9e3e762da95615d6759d2d837226917089ffffd7. `gitPullIfNeeded` no longer logs that it reads `lyra.yml`

File: webapp/src/store/Store.ts
Comment (WULCAN) [reply webapp/src/store/Store.ts:4]: Yes, I did not consider that one of the mappings are for all the projects inside a single repository, and the other mapping is for all projects configured in the Lyra instance.

File: webapp/src/app/api/translations/[projectName]/[lang]/[msgId]/route.ts
Comment (amerharb) [reply webapp/src/app/api/translations/[projectName]/[lang]/[msgId]/route.ts:40]: I have added this todo:
  // TODO: include getProjectConfig & readFromDir in a try/catch block and check for error to return a certain 500 error

File: webapp/src/utils/config.ts
Comment (WULCAN) [reply webapp/src/utils/config.ts]: With my first suggested resolution, we would lose the ability to run Lyra for a subset of the projects in a repository.

The second proposal is still viable, but still too big to include in this merge request.

File: webapp/src/utils/config.spec.ts
Comment (WULCAN) [reply webapp/src/utils/config.spec.ts]: Yes, the conclusion was that expect.assertions is still valuable as it helps if we forget to await or return properly.

File: webapp/src/utils/config.ts
Comment (WULCAN) [reply webapp/src/utils/config.ts]: Resolved in 90ff29e31c6c70ab7884f44e8bbab08591068e28 by extending the TODO comment.

File: webapp/src/app/[projectName]/[lang]/page.tsx
Comment (WULCAN) [reply webapp/src/app/[projectName]/[lang]/page.tsx]: Resoled on fc01da3af985a05d10d225215c9e1afc82f95586 by moving the destructuring to a statement inside the function.

File: webapp/src/utils/lyraConfig.ts
Comment (amerharb) [reply webapp/src/utils/lyraConfig.ts]: fixed now both are normlized

Comment (WULCAN) [reply webapp/src/utils/lyraConfig.ts]: Resolved on 30a65c9945179377cde9acbdf0a04b193c6a951d by normalizing ServerConfig too.

File: webapp/src/app/[projectName]/[lang]/page.tsx
Comment (amerharb) [reply webapp/src/app/[projectName]/[lang]/page.tsx]: fixed by making a local type Context

File: webapp/src/app/[projectName]/[lang]/page.tsx
Comment (amerharb) [reply webapp/src/app/[projectName]/[lang]/page.tsx]: I leave this for now when a good FE design will happend

File: webapp/src/store/Store.spec.ts
Comment (amerharb) [reply webapp/src/store/Store.spec.ts]: Agree, for now I will change it basicly to toBe() then later when we have a redis or similar solution we will rewrite the unit test

Comment (WULCAN) [reply webapp/src/store/Store.spec.ts]: Resolved in 0511e6019a9787bd5d1476499e65187288869a3b by using `toBe` instead of `toEqual`.

File: webapp/src/utils/config.ts
Comment (amerharb) [reply webapp/src/utils/config.ts]: host removed

File: webapp/src/app/[projectName]/[lang]/page.tsx
Comment (amerharb) [reply webapp/src/app/[projectName]/[lang]/page.tsx]: anyway reverted

File: webapp/src/app/api/pull-request/[projectName]/route.ts
Comment (amerharb) [reply webapp/src/app/api/pull-request/[projectName]/route.ts:41]: update msg

File: webapp/README.md
Comment (amerharb) [reply webapp/README.md:7]: updated

File: webapp/src/utils/config.ts
Comment (amerharb) [reply webapp/src/utils/config.ts]: jsdoc added

File: webapp/src/utils/config.ts
Comment (amerharb) [reply webapp/src/utils/config.ts]: After discussion with @WULCAN and @richardolsson 
this setup currently of project.yaml is temporarily and as you mention we could do in separate PR or maybe a better solution will be in place

File: webapp/src/utils/config.spec.ts
Comment (amerharb) [reply webapp/src/utils/config.spec.ts]: I think the solution  is anti pattern, since we are going to make a private constructor a public one just to make it testable.
I think it is okay for now to have a wide scope unit test rather than do that

File: webapp/src/Cache.ts
Comment (amerharb) [reply webapp/src/Cache.ts]: That is absolutely a good argument, we did not think or design for configuration changes during update the translation.

I suggest to convert this into an issue to investigate

File: webapp/src/utils/config.ts
Comment (WULCAN) [reply webapp/src/utils/config.ts]: Resolved on 40dac12e036efb0ce4cd83ed1d95ac6caa7ffa02 by removing host.

File: webapp/src/app/api/pull-request/[projectName]/route.ts
Comment (WULCAN) [inline webapp/src/app/api/pull-request/[projectName]/route.ts:41]: Good catch with the risk of multiple Lyra projects using the same local clone. As a future improvement, we should reject such configurations when starting Lyra.

File: webapp/README.md
Comment (WULCAN) [reply webapp/README.md]: Resolved on 5133faf52783059fa0882bebfcb18d201a1034e8 with a new sentence.

File: webapp/src/utils/config.ts
Comment (WULCAN) [reply webapp/src/utils/config.ts]: On 5133faf52783059fa0882bebfcb18d201a1034e8, README documents the paths very well.

Comment (WULCAN) [reply webapp/src/utils/config.ts]: Resolved on b4f5dd6af651cf7449ee0552c758bc6b6de72d26 with jsdoc.

**Review by WULCAN — APPROVED** [review]
Excellent work @amerharb! I have reviewed all the commits up to and including 6faae5c43acd17eb48b557726a2f28755106ddc8 and I approve of merging them into main.

**Comment by amerharb** [comment]
> Excellent work @amerharb! I have reviewed all the commits up to and including [6faae5c](https://github.com/zetkin/lyra/commit/6faae5c43acd17eb48b557726a2f28755106ddc8) and I approve of merging them into main.

Greate @WULCAN, I can see the check of approve, but still I can't merge it, I think you need to have a role with write access first.

@richardolsson can you check this?

**Review by richardolsson — APPROVED** [review]
I'm not sure why the review from @WULCAN doesn't enable merging, but here's my approval too. Nice work!

**Pull request merged**
